File: NBOv1-Modules/NuSoft004/UILayer/Transaksi/UI_SaldoAwalDialog.vb
Imports DevExpress.Xpo
Imports NuSoft.Core.SSystem
Imports DevExpress.Data.Filtering

Public Class UI_SaldoAwalDialog
  Private instance As Persistent.StockMain
  Private xpSettingRegional As XPCollection(Of NuSoft001.Persistent.SettingRegional)
  Private defaultRegional As NPO.Modules.ModSys.Regional
  Private setting As NuSoft001.Logic.FinaSetting
  Private sumber As Persistent.SumberDataCollection
  Sub New()
    ' This call is required by the designer.
    InitializeComponent()
    ' Add any initialization after the InitializeComponent() call.
    AutoCloseOnSave = False
  End Sub
  Overrides Sub LoadBeforeInitialize()
    If My.Computer.FileSystem.FileExists(Logic.Umum.xLokasiLayout("BSTransaksi.xml")) = True Then
      lytTransaksiSA.RestoreLayoutFromXml(Logic.Umum.xLokasiLayout("BSTransaksi.xml"))
    End If
    sumber = New Persistent.SumberDataCollection(session)
    txtBarang.Properties.DataSource = New XPCollection(Of Persistent.Barang)(session)
    txtRegional.Properties.DataSource = UserInfo.GetUserRegional(session, NamaDatabase)
    txtSatuan.Properties.DataSource = New XPCollection(Of Persistent.Satuan)(session)
    txtGudang.Properties.DataSource = New XPCollection(Of Persistent.Gudang)(session)
    txtGudang.EditValue = Core.Win.Component.XtraLookUpEdit.GetFirstValue(txtGudang)
    txtKaryawan.Properties.DataSource = New XPCollection(Of NuSoft001.Persistent.Kontak)(session)
    defaultRegional = UserInfo.GetDefaultRegional(session)
    If Not defaultRegional Is Nothing Then
      txtRegional.EditValue = defaultRegional
    Else
      txtRegional.EditValue = Core.Win.Component.XtraLookUpEdit.GetFirstValue(txtRegional)
    End If
  End Sub
  Overrides Sub InitializeData()
    setting = New NuSoft001.Logic.FinaSetting(session)
    If Tipe = InputType.Tambah Then
      Me.Text = "Saldo Awal: Tambah"
      instance = New Persistent.StockMain(session)
      instance.Gl = New Persistent.GlMain(session) With {.Number = New Logic.NumberClass(session) With {.FormatNomor = setting.FormatNomor}}
      txtTanggal.EditValue = Now
      instance.Tanggal = txtTanggal.DateTime
      instance.Regional = CType(txtRegional.EditValue, NPO.Modules.ModSys.Regional)
      ChangeCode()
      txtHarga.EditValue = 0.0
      txtQty.EditValue = 0.0
      LoadDefault()
    Else
      instance = session.GetObjectByKey(Of Persistent.StockMain)(CType(IdToEdit, Int64))
      txtTanggal.EditValue = instance.Tanggal
      txtNoTransaksi.EditValue = instance.Kode
      txtRegional.EditValue = instance.Regional
      txtGudang.EditValue = instance.Gudang
      txtBarang.EditValue = instance.Detail(0).Barang
      txtQty.EditValue = instance.Detail(0).QtyIn
      txtSatuan.EditValue = instance.Detail(0).Satuan
      txtHarga.EditValue = instance.Detail(0).HargaIn
      txtKaryawan.EditValue = instance.KaryawanKirim
      Me.Text = "Saldo Awal Edit - " & instance.Kode
      'cek dulu apakah sudah ada HPPnya, jika sudah ada maka dilock
      Dim xHPP As New XPCollection(Of Persistent.StockDetailHPPReff)(session, New BinaryOperator("DetailIN.Id", instance.Detail(0).Id, BinaryOperatorType.Equal))
      If xHPP.Count > 0 Then
        EnableVisibleSaveButton(False, False)
      End If
    End If
    SetNomor()
  End Sub
  Overrides Sub LoadAfterInitialize()
    txtQty.Properties.Mask.EditMask = setting.NumericFormatToString
    txtHarga.Properties.Mask.EditMask = setting.NumericFormatToString
  End Sub
  Overrides Sub SimpanData()
    If Tipe = InputType.Tambah Then
      ChangeCode()
      For pi = 0 To instance.Detail.Count - 1
        instance.Detail(pi).Delete()
      Next
    End If
    Dim MataUangDefault As NuSoft001.Persistent.MataUang, zNilai As Double
    MataUangDefault = session.GetObjectByKey(Of NuSoft001.Persistent.MataUang)(Convert.ToInt32(setting.MultiMataUangDefault))
    instance.Kode = CType(txtNoTransaksi.EditValue, String)
    instance.Tanggal = CType(txtTanggal.EditValue, Date)
    instance.Regional = CType(txtRegional.EditValue, NPO.Modules.ModSys.Regional)
    instance.Gudang = CType(txtGudang.EditValue, Persistent.Gudang)
    instance.KaryawanKirim = CType(txtKaryawan.EditValue, NuSoft001.Persistent.Kontak)
    instance.MataUang = MataUangDefault
    instance.Kurs = 1
    instance.StatusTransaksi = NuSoft001.Persistent.StatusTransaksiEnum.Posting
    If Tipe = InputType.Tambah Then
      Dim zData As Persistent.StockDetail
      zData = New Persistent.StockDetail(session)
      zData.Main = instance
      zData.Barang = CType(txtBarang.EditValue, Persistent.Barang)
      zData.QtyIn = CType(txtQty.EditValue, Double)
      zData.Satuan = CType(txtSatuan.EditValue, Persistent.Satuan)
      zData.HargaIn = CType(txtHarga.EditValue, Double)
      instance.Detail.Add(zData)
    Else
      instance.Detail(0).Barang = CType(txtBarang.EditValue, Persistent.Barang)
      instance.Detail(0).QtyIn = CType(txtQty.EditValue, Double)
      instance.Detail(0).Satuan = CType(txtSatuan.EditValue, Persistent.Satuan)
      instance.Detail(0).HargaIn = CType(txtHarga.EditValue, Double)
    End If
    zNilai = instance.Detail(0).QtyIn * instance.Detail(0).HargaIn
    '==== jurnalnya
    instance.Gl.Sumber = instance.Sumber
    instance.Gl.Tanggal = instance.Tanggal
    instance.Gl.Regional = CType(txtRegional.EditValue, NPO.Modules.ModSys.Regional)
    instance.Gl.Kode = instance.Kode
    instance.Gl.ModuleId = NPO.Modules.ModuleId.Purchase
    instance.Gl.MataUang = MataUangDefault.Simbol
    instance.Gl.Kurs = 1
    instance.Gl.Kontak = CType(txtKaryawan.EditValue, NuSoft001.Persistent.Kontak)
    instance.Gl.KontakNama = CType(txtKaryawan.EditValue, NuSoft001.Persistent.Kontak).Nama
    instance.Gl.Uraian = "Saldo Awal Persediaan"
    SimpanGL(CType(txtBarang.EditValue, Persistent.Barang).CoaStock, 1, zNilai, 0, 0, 0)
    SimpanGL(session.GetObjectByKey(Of NuSoft001.Persistent.Coa)(Convert.ToInt32(setting.PenampungSaldoAwal)), 2, 0, zNilai, 0, 0, "")

    instance.Save()
    session.CommitChanges()
  End Sub
  Private Sub SimpanGL(Akun As NuSoft001.Persistent.Coa, Urutan As Byte, Debit As Double, Kredit As Double, DebitValas As Double, KreditValas As Double, Optional Catatan As String = "")
    Dim item As Persistent.GlMainDetail
    If instance.Gl.Detail.ToList.Exists(Function(m) m.Urutan = Urutan) Then
      item = instance.Gl.Detail.Single(Function(m) m.Urutan = Urutan)
      item.Akun = Akun
      item.MataUang = item.Akun.MataUang
      item.Main = instance.Gl
      item.Urutan = Urutan
      item.Debit = Debit
      item.Kredit = Kredit
      item.Keterangan = Catatan
      item.DebitValas = 0
      item.KreditValas = 0
      item.Kurs = 1
    Else
      item = New Persistent.GlMainDetail(session)

      item.Akun = Akun
      item.MataUang = item.Akun.MataUang
      item.Main = instance.Gl
      item.Urutan = Urutan
      item.Debit = Debit
      item.Kredit = Kredit
      item.Keterangan = Catatan
      item.DebitValas = 0
      item.KreditValas = 0
      item.Kurs = 1
      instance.Gl.Detail.Add(item)
    End If
  End Sub
  Overrides Sub ErrorSimpan(ByVal ex As Utils.Exception)
    ex.ShowWinMessageBox()
    Select Case ex.ErrorNumber
      Case -1 : txtBarang.Focus()
      Case -2 : txtGudang.Focus()
      Case -3 : txtQty.Focus()
      Case Else
    End Select
  End Sub
  Private Sub txtGudang_KeyDown(sender As Object, e As System.Windows.Forms.KeyEventArgs) Handles txtGudang.KeyDown
    If e.KeyCode = Windows.Forms.Keys.Delete Then
      txtGudang.EditValue = Nothing
    End If
  End Sub
  Private Sub txtBarang_EditValueChanging(sender As Object, e As DevExpress.XtraEditors.Controls.ChangingEventArgs) Handles txtBarang.EditValueChanging
    If Not e.NewValue Is Nothing Then
      txtSatuan.EditValue = CType(e.NewValue, Persistent.Barang).Satuan
      txtHarga.EditValue = CType(e.NewValue, Persistent.Barang).LastPurchase
    End If
  End Sub
  Private Sub txtBarang_KeyDown(sender As Object, e As System.Windows.Forms.KeyEventArgs) Handles txtBarang.KeyDown
    If e.KeyCode = Windows.Forms.Keys.Delete Then
      txtBarang.EditValue = Nothing
    End If
  End Sub
  Sub LoadDefault()
    If xpSettingRegional Is Nothing Then
      xpSettingRegional = New XPCollection(Of NuSoft001.Persistent.SettingRegional)(session)
    End If
    defaultRegional = UserInfo.GetDefaultRegional(session)
    txtTanggal.EditValue = New Date(setting.AwalPeriodeTahun, setting.AwalPeriodeBulan, 1)
    If Not defaultRegional Is Nothing Then
      txtRegional.EditValue = defaultRegional
    Else
      txtRegional.EditValue = Core.Win.Component.XtraLookUpEdit.GetFirstValue(txtRegional)
    End If
  End Sub
  Private Sub ChangeCode()
    If txtTanggal.EditValue Is Nothing OrElse instance.Tanggal = Nothing Then
      Return
    End If
    If setting.FormatNomor.Contains(NuSoft001.My.Resources.FormatRegional) AndAlso (txtRegional.EditValue Is Nothing OrElse instance.Regional Is Nothing) Then
      Return
    End If
    If setting.FormatNomor.Contains(NuSoft001.My.Resources.FormatSumber) AndAlso instance.Sumber Is Nothing Then
      Return
    End If

    Dim nomor As Core.Modules.NumberResult = instance.Number.GetNewNumber
    txtNoTransaksi.Text = nomor.Kode
  End Sub
  Private Sub txtRegional_EditValueChanging(sender As Object, e As DevExpress.XtraEditors.Controls.ChangingEventArgs) Handles txtRegional.EditValueChanging
    Dim str As String = "Edit regional transaksi akan mengganti nomor bukti." & vbCrLf & "Apakah anda ingin meneruskan ?"
    Try
      If Tipe = InputType.Edit Then
        If Not instance.Regional Is e.NewValue Then
          If DevExpress.XtraEditors.XtraMessageBox.Show(str, "Ganti Regional", Windows.Forms.MessageBoxButtons.YesNo, Windows.Forms.MessageBoxIcon.Question) = Windows.Forms.DialogResult.Yes Then
            instance.Regional = CType(e.NewValue, NPO.Modules.ModSys.Regional)
            ChangeCode()
          Else
            e.Cancel = True
          End If
        Else
          instance.Regional = CType(e.NewValue, NPO.Modules.ModSys.Regional)
          ChangeCode()
        End If
      ElseIf Tipe = InputType.Tambah Then
        instance.Regional = CType(e.NewValue, NPO.Modules.ModSys.Regional)
        ChangeCode()
      End If
    Catch ex As Exception
      If ex.Message.Contains("Object reference not set to an instance") Then
      Else
        DevExpress.XtraEditors.XtraMessageBox.Show(ex.Message, "Ganti Regional", Windows.Forms.MessageBoxButtons.OK, Windows.Forms.MessageBoxIcon.Error)
      End If
    End Try
  End Sub
  Private Sub txtTanggal_DateTimeChanged(sender As Object, e As EventArgs) Handles txtTanggal.DateTimeChanged
    Dim str As String = "Edit tanggal dengan bulan dan tahun berbeda akan mengganti nomor bukti." & vbCrLf & "Apakah anda ingin meneruskan ?"
    Try
      If Tipe = InputType.Edit Then
        If instance.Tanggal.Month <> CDate(txtTanggal.EditValue).Month OrElse instance.Tanggal.Year <> CDate(txtTanggal.EditValue).Year Then
          If DevExpress.XtraEditors.XtraMessageBox.Show(str, "Ganti Tanggal", Windows.Forms.MessageBoxButtons.YesNo, Windows.Forms.MessageBoxIcon.Question) = Windows.Forms.DialogResult.Yes Then
            instance.Tanggal = txtTanggal.DateTime
            ChangeCode()
          Else
            txtTanggal.EditValue = instance.Tanggal
          End If
        Else
          instance.Tanggal = txtTanggal.DateTime
        End If
      ElseIf Tipe = InputType.Tambah Then
        instance.Tanggal = txtTanggal.DateTime
        ChangeCode()
      End If
    Catch ex As Exception
      If ex.Message.Contains("Object reference not set to an instance") Then
      Else
        DevExpress.XtraEditors.XtraMessageBox.Show(ex.Message, Me.Text, Windows.Forms.MessageBoxButtons.OK, Windows.Forms.MessageBoxIcon.Error)
      End If
    End Try
  End Sub
  Private Sub SetNomor()
    instance.Sumber = sumber.GetObject(Persistent.SumberDataJenis.BS)
    instance.Number = New Logic.NumberClass(session) With {.FormatNomor = setting.FormatNomor}
  End Sub
End Class

File: NBOv1-Modules/Nusoft012/UI/Transaksi/UI_SaldoAwalDialog.cs
using DevExpress.Xpo;
using NuSoft.Core.Win.Forms;
using NuSoft.NUI.Win.Forms.Modules.NuSoft007.Services;
using NuSoft.NUI.Win.Forms.Modules.NuSoft012.Persistent;
using NuSoft.NUI.Win.Forms.Modules.NuSoft012.Services;
using System;
using System.Collections.Generic;
using System.Linq;

namespace NuSoft.NUI.Win.Forms.Modules.NuSoft012.UI.Transaksi {
	public partial class UI_SaldoAwalDialog : InputDialog {
		public UI_SaldoAwalDialog() {
			InitializeComponent();

			txtTanggal.DateTimeChanged += new EventHandler(TanggalChanged);
			txtSaldoPiutang.ValueChanged += new EventHandler(PiutangChanged);
			AutoCloseOnSave = true;
		}
		Invoice originalEdit;

		private void TanggalChanged(object sender, EventArgs e) {
			txtJatuhTempo.Properties.MinValue = txtTanggal.DateTime;
		}
		private void PiutangChanged(object sender, EventArgs e) {
			txtOmzet.Properties.MinValue = txtSaldoPiutang.Value;
			txtOmzet.Properties.MaxValue = decimal.MaxValue;
		}

		private void CheckDisableControl(Invoice obj) {
			bool disable;
			try { InvoiceService.CheckIsInUse(session, obj); disable = false; }
			catch { disable = true; }

			if (disable) {
				txtWilayah.Enabled = false;
				txtTipeInvoice.Enabled = false;
				txtNomorInvoice.Enabled = false;
				txtTanggal.Enabled = false;
				txtJatuhTempo.Enabled = false;
				txtSales.Enabled = false;
				txtPemasang.Enabled = false;
				txtSaldoPiutang.Enabled = false;
				txtOmzet.Enabled = false;
				txtCashbackPersen.Enabled = false;
				txtCashbackNominal.Enabled = false;
				txtKeterangan.Enabled = false;
				txtKomisiPersen.Enabled = false;
				txtKomisiNominal.Enabled = false;

				EnableVisibleSaveButton(false, true);
			}
		}

		public override void LoadBeforeInitialize() {
			txtWilayah.Properties.DataSource = new XPCollection<Wilayah>(session).ToList();

			txtTipeInvoice.Properties.DataSource = new XPCollection<TipeInvoice>(session);
			txtNomorInvoice.Properties.MaxLength = 30;
			txtSales.Properties.DataSource = new XPQuery<Sales>(session).Where(w => w.Aktif).ToList();
			txtPemasang.Properties.MaxLength = 100;
			txtKeterangan.Properties.MaxLength = 255;

			txtWilayah.EditValue = ((List<Wilayah>)txtWilayah.Properties.DataSource)[0];
			txtTipeInvoice.EditValue = ((XPCollection<TipeInvoice>)txtTipeInvoice.Properties.DataSource)[0];
			txtSales.EditValue = ((List<Sales>)txtSales.Properties.DataSource)[0];
		}
		public override void InitializeData() {
			if (Tipe == InputType.Tambah) {
				Text = "Saldo Awal Piutang : Tambah";
				txtNomorInvoice.Text = "";
				txtTanggal.DateTime = DateTime.Now;
				txtJatuhTempo.DateTime = DateTime.Now;
				txtPemasang.Text = "";
				txtSaldoPiutang.Value = 0;
				txtOmzet.Value = 0;
				txtKomisiPersen.Value = 0;
				txtKomisiNominal.Value = 0;
				txtCashbackPersen.Value = 0;
				txtKeterangan.Text = "";
			}
			else {
				originalEdit = session.GetObjectByKey<Invoice>(Convert.ToInt64(IdToEdit));
				Text = "Saldo Awal Piutang : Edit - " + originalEdit.NoInvoice;
				txtWilayah.EditValue = originalEdit.Wilayah;
				txtTipeInvoice.EditValue = originalEdit.TipeInvoice;
				txtNomorInvoice.Text = originalEdit.NoInvoice;
				txtTanggal.DateTime = originalEdit.TanggalOmzet;
				txtJatuhTempo.DateTime = originalEdit.TanggalJatuhTempo;
				txtSales.EditValue = originalEdit.Sales;
				txtPemasang.Text = originalEdit.InvoiceNama;
				txtSaldoPiutang.Value = originalEdit.Total;
				txtOmzet.Value = originalEdit.Netto;
				txtCashbackPersen.Value = originalEdit.CashbackPersen;
				txtCashbackNominal.Value = originalEdit.CashbackNominal;
				txtKeterangan.Text = originalEdit.Keterangan;
				txtKomisiPersen.Value = originalEdit.KomisiPersen;
				txtKomisiNominal.Value = originalEdit.KomisiNominal;
				CheckDisableControl(originalEdit);
			}
			txtWilayah.Focus();
		}
		public override void SimpanData() {
			Invoice instance;
			if (Tipe == InputType.Tambah) instance = new Invoice(session);
			else instance = session.GetObjectByKey<Invoice>(Convert.ToInt64(IdToEdit));
			var service = new InvoiceService(session, originalEdit);

			instance.Wilayah = txtWilayah.EditValue == null ? null : (Wilayah)txtWilayah.EditValue;
			instance.TipeInvoice = txtTipeInvoice.EditValue == null ? null : (TipeInvoice)txtTipeInvoice.EditValue;
			instance.NoInvoice = txtNomorInvoice.Text;
			instance.TanggalOmzet = txtTanggal.DateTime;
			instance.TanggalJatuhTempo = txtJatuhTempo.DateTime;
			instance.Sales = txtSales.EditValue == null ? null : (Sales)txtSales.EditValue;
			instance.InvoiceNama = txtPemasang.Text;
			instance.Total = txtSaldoPiutang.Value;
			instance.Netto = txtOmzet.Value;
			instance.KomisiPersen = txtKomisiPersen.Value;
			instance.KomisiNominal = txtKomisiNominal.Value;
			instance.CashbackPersen = txtCashbackPersen.Value;
			instance.CashbackNominal = txtCashbackNominal.Value;
			instance.Keterangan = txtKeterangan.Text;

			instance.SaldoAwal = true;
			instance.TanggalInvoice = txtTanggal.DateTime;
			if (instance.Wilayah != null) instance.InvoiceAlamat = instance.Wilayah.Nama;
			instance.InvoiceNPWP = PPnKeluaranService.GetDefaultNPWP();
			instance.Bruto = 0;
			instance.DiskonPersen = 0;
			instance.DiskonNominal = 0;
			instance.HargaTermasukPajak = false;
			instance.PajakPersen = 0;
			instance.PajakNominal = 0;

			service.Save(instance);
		}
		public override void ErrorSimpan(Utils.Exception ex) {
			ex.ShowWinMessageBox();
			switch (ex.ErrorNumber) {
				case -1:
				case -2: txtTanggal.Focus(); break;
				case -3: txtJatuhTempo.Focus(); break;
				case -4: txtWilayah.Focus(); break;
				case -5: txtSales.Focus(); break;
				case -6: txtTipeInvoice.Focus(); break;
				case -7: txtPemasang.Focus(); break;
				case -8: txtWilayah.Focus(); break;
				case -9: txtPemasang.Focus(); break;
				case -10: txtOmzet.Focus(); break;
				case -11: txtSaldoPiutang.Focus(); break;
				case -99: txtNomorInvoice.Focus(); break;
			}
		}
	}
}
